### PR TITLE
[translation] Post: Introduction à React

### PR DESCRIPTION
**Pull request by skinnyfoetusboy**
Okay, I tried to do this quickly just to give a taste of what it could look like.
This is in no way final or anything, just a quick preview.

I tried to keep @bloodyowl's writing style while not making it too literal a translation (which was kinda easy; boy how many anglicisms do you use?)

### CODE REVIEW

**Comment by bloodyowl** [comment]
je review ça tout à l'heure

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+It's been quite some time since ReactJS has been open-sourced by facebook.
+Let's take it as an opportunity to introduce the last-born child of the MV* trend.
+
+## Different takes on UI
+
+### Backbone
+
+[Backbone](http://backbonejs.org) brings a simplification in events declaration,
+but doesn't indulge quite that much in rendering.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-doesn't indulge quite
+doesn't handle
```

Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-in
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+Making the developer the sole master of `Backbone.View`
+is positive for several kinds of projects, but makes
+DOM manipulation tedious.
+
+On the subject of data-binding, `Backbone` doesn't actually come
+with any in-house solutions and you're stuck with using mixins to
+implement it without resorting to gnawing your own legs off.
+
+Moreover, it's usually a templating engine ala Handlebars or Jade
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-ala
+*à la*
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+that creates the original view. Thus, you find yourself with some sort
+of `this.$el.html(this.template(data))` in your `render()` method which
+will leave you with a few UX aberrations like your pictures re-rendering
+themselves or videos playing back from the start.
+
+In a nutshell, Backbone is pretty useful to create some sort of
+structure your code, but it is absolutely lackluster in
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-create some sort of structure your code
+help you structure your code
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+It also has strong design patterns and contrary to what people commonly
+think, it's actually pretty straightforwad.
+Subviews are also pretty easy to use thanks to `{{outlet}}`.
+
+In another words, Ember and Angular both come bundled with real
+solutions for UI management. But they also share the same way to
+do it :
+
+- _rendering_ (creating and inserting DOM nodes) once
+- updating the bindings
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]: that is no longer true for ember, but we'll keep the original date for people not to get confused

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+### React
+
+[React](http://facebook.github.io/react/) doesn't care what other people do.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-doesn't care what other people do
+brings a completely different approach
```

File: content/en/posts/js/introduction-a-reactjs/index.md (hunk 4 above)
Comment (skinnyfoetusboy) [inline content/en/posts/js/introduction-a-reactjs/index.md]: Maybe a disclaimer at the top of the article too?

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+It has one simple way of going about stuff :
+holy crap, managing the state of the DOM _sucks_
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-holy crap, managing the state of the DOM _sucks_
+managing the DOM state is a bloody mess
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+Okay, let's just say we call `.render()` everytime anything changes.
+That might sound _real_ dumb but there's a gotcha about it.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-That might sound _real_ dumb but there's a gotcha about it
+Sounds dumb, right? Actually not that much.
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+React has a __virtual DOM__ (nothing to do with facebook's purchase of Oculus VR),
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-React has a __virtual DOM__ (nothing to do with facebook's purchase of Oculus VR),
+React implements a __virtual DOM__,
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+a blazing-fast internal representation of the DOM. It comes with its own events system
+which allows React to bring the capturing phase to browsers that don't have
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]: actually that was my mistake.

```diff
-the capturing phase
+a consistent event system
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+`EventTarget` (looking at you, IE8).
+
+The `render` method returns objects corresponding to the internal representation
+of the virtual DOM.
+
+React classes are defined by their `state`.
+When creating a class, you specify a `getInitialState` method which
+will return the initial state (duh!).
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-(duh!)
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+After that, the only way to modify the state is to pass which values
+of the state have changed to `this.setState` so that it can update
+the DOM.
+
+A React class takes properties when it's instanciated : they're called `props`.
+They shouldn't be confused with the `state`, its content should only be manipulated
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-its
+their
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+from _outside_ of the the class (even though it can still get default values by
+specifying a `getDefaultProps` method which will return them).
+
+The `state`, however, shall only be modified by the class's own methods.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-shall
+should
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+The main advantage to this is that you're always sure, thanks to the systematic
+call to `render`, that your React component will look and act like you expect it
+to.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-that your React component will look and act like you expect it to
+that your React component will have the expected representation.
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+Another advantage to React is its in-house diff algorithm.
+The virtual DOM gets diffed against the previous one (which is the one you see),
+and React works out which operations it needs to execute in order to update the 
+DOM.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-which operations it needs to execute
+the smallest amount of operations needed
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+
+This solves a few issues like keeping track of the caret's position in a text
+field that uses two-way data-binding : since the algorithm doesn't see any reason
+to update it, the text field doesn't get re-rendered and you keep the focus.
+In the same way, if you've got a gif looping somewhere in your page, it will keep
+playing as it normally would.
+
+React is made to work with JSX (that's JavaScript XML), a superset of JS which
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-React is made to work with JSX
+React plays very well with JSX
```

Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]: also maybe remove `(that's JavaScript XML)`, redundant with the following `XML syntax`

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+allows you to write your templates with an XML syntax (see example below),
+alowing even beginners to get stuff done with it quickly.
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-even
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+```javascript
+var View = React.createClass({
+  getInitialState : function(){
+    // initial state
+    return {
+      checked : false
+    }
+  },
+  getDefaultProps : function(){
+    // if `this.props.label` isn't defined, then `"?"` will be displayed
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-will be displayed
+will be used
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+      <div className={classes}>
+        {/* our binding, simple as that*/}
+        <input checked={this.state.checked} type="checkbox" onChange={this.toggle} />
+        {this.props.label}
+      </div>
+    )
+  }
+})
+
+// let's mount the component and pass the label to it
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-let's
+we
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+var view = React.render(<View label="helloworld" />, document.getElementById("id"))
+// and voilà!
+view.toggle()
+```
+
+## Sum up of React's advantages
+
+React has this figured out:
+
+- DOM i/o is slow and not accessing it too often is crucial;
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-DOM i/o is slow and not accessing it too often is crucial;
+the DOM is slow, particularly its writes, and limiting interactions with it is crucial;
```

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+- always thinking about the state the DOM has at every moment is not what we should
```
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]:
```diff
-always thinking about the state the DOM has at every moment is not what we should
+always thinking about the DOM the state has at any point in time is not what we should
```

File: content/en/posts/js/introduction-a-reactjs/index.md (hunk 4 above)
Comment (bloodyowl) [inline content/en/posts/js/introduction-a-reactjs/index.md]: I'll do this after, and update both articles.

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 21]
```diff
@@ -0,0 +1,176 @@
+---
+date: "2014-06-04"
+route: /posts/js/introduction-a-reactjs/
+title: Introducing ReactJS
+tags:
+  - javascript
+  - reactjs
+authors:
+  - bloodyowl
+  - skinnyfoetusboy
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.md]: I would go for adding a "translators" field. What do you think ?

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 22]
```diff
@@ -0,0 +1,176 @@
+---
+date: "2014-06-04"
+route: /posts/js/introduction-a-reactjs/
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.md]: please rename the file and remove this field (related to #380)

File: content/en/posts/js/introduction-a-reactjs/index.md [hunk 23]
```diff
@@ -0,0 +1,176 @@
+---
+date: "2014-06-04"
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.md]: Can you add as a first field `lang: en` ?

File: content/en/posts/js/introduction-a-reactjs/index.md (hunk 21 above)
Comment (skinnyfoetusboy) [inline content/en/posts/js/introduction-a-reactjs/index.md]: I would think it's a pretty good idea. 
I only added myself as an author for lack of a better place.

File: content/en/posts/js/introduction-a-reactjs/index.md (hunk 22 above)
Comment (skinnyfoetusboy) [inline content/en/posts/js/introduction-a-reactjs/index.md]: Renaming it to what?
Something like "index.en.md"?

File: content/en/posts/js/introduction-a-reactjs/index.md (hunk 20 above)
Comment (skinnyfoetusboy) [inline content/en/posts/js/introduction-a-reactjs/index.md]: "the DOM the state has" doesn't quite compute in my head

File: content/en/posts/js/introduction-a-reactjs/index.md (hunk 22 above)
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.md]: I would go for "content/en/posts/js/introduction-TO-reactjs/index.md" (in lowercase)

File: content/en/posts/js/introduction-a-reactjs/index.en.md [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+lang: en
+date: "2014-06-04"
+title: Introducing ReactJS
+tags:
+  - javascript
+  - reactjs
+authors:
+  - bloodyowl
+translators:
+  - skinnyfoetusboy
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.en.md]: :+1:

File: content/en/posts/js/introduction-a-reactjs/index.en.md [hunk 25]
```diff
@@ -0,0 +1,175 @@
+---
+lang: en
+date: "2014-06-04"
+title: Introducing ReactJS
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.en.md]: Please update file name according to this title.

File: content/en/posts/js/introduction-a-reactjs/index.en.md [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+tags:
+  - javascript
+  - reactjs
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.en.md]: react has tag, not reactjs

File: content/en/posts/js/introduction-a-reactjs/index.en.md [hunk 27]
```diff
@@ -0,0 +1,175 @@
+---
+lang: en
```
Comment (MoOx) [inline content/en/posts/js/introduction-a-reactjs/index.en.md]: Sorry but you can remove this lang meta, I will handle it from the filename :)

**Comment by MoOx** [comment]
Nice update. If you have some times, maybe squash to one commit to avoid the weird merge. Otherwise I might do it when merging.
I put the PR in blocked state until we get an english homepage :)

**Comment by MoOx** [comment]
@skinnyfoetusboy can you double check you have used the same mail you use for your github account? It seems you are not according to this line https://travis-ci.org/putaindecode/putaindecode.fr/builds/63131895#L530

**Comment by Nyalab** [comment]
pretty good, even if it is a bit late to introduce to react in english by now

**Comment by MoOx** [comment]
the idea is to have content for .io :)

**Comment by MoOx** [comment]
Also: it's never too late

**Comment by skinnyfoetusboy** [comment]
@MoOx It seems like my git account wasn't set up at the time I made this contribution.
What's the way to go in order to fix this?

**Comment by MoOx** [comment]
Probably something like `git rebase --interactive` or something to edit all commit authors when accurate. Maybe just rebase + squash, it seems 2 commits on 5 are not ok.

**Comment by Nyalab** [comment]
tu t'en sortiras à base de `git rebase -i HEAD~nbcommits` et de `git commit --amend --author="Author <mail@author>"`

**Comment by MoOx** [comment]
This post will be alive with when #487 will be closed :)

**Pull request merged**